### PR TITLE
`sub` directory

### PR DESCRIPTION
**Pull request by EECOLOR**
This pull request follows from a discussion about directories.

The template contains a `components` directory. It felt a bit weird for this directory to have files other than actual React components (or their styles).

We noticed the following:
- `components` is the opposite for `pages`
- `domain` is the opposite for `machinery`

We wanted a directory for things that a not `pages` and not `machinery`. The problem was that leaving out a directory would hinder the organisation of files. But the `components`/`domain` part should be implicit by the fact that the file was not in `machinery` or `pages`.

This drove us to the following:
- `sub` for both `domain` and `components`
- Root slash imports first search in `sub`

In this pull request I have adjusted the template to reflect this.

When reviewing it is important to think about backward compatibility. Does this change break any of our existing websites?

### CODE REVIEW

**Comment by EECOLOR** [comment]
Note to self: add `sub` to the `example` test application and test `start` / `build`, etc..

**Comment by peeke** [comment]
While I get the gist of it, it would be useful to understand what the discussion was about. What I'm assuming now is that at some point a machinery folder was added to a domain folder, but to have a machinery folder inside of the components folder didn't make sense.

The name `sub` isn't very meaningful to me. Sub- what? Maybe we should take some time to find some alternatives. 

Since a component can be part of a domain, but domain doesn't communicate it will only include components, did you consider other solutions? Like making `domain` the top level folder, including a domain folder with a components folder: `/domain/home/components/Component`? Or dropping the `components` folder so you have `/domain/home/Component`?

I have tested in a few projects, all of which function fine, so the mechanism seems to be working :+1:

File: library/eslint-resolver-plugins/absolute-path-resolver-plugin.js [hunk 1]
```diff
@@ -4,12 +4,14 @@ const path = require('path')
 
 const modifiers = [
   function removeQueryString(source, file, config) {
-    const [newSource] = source.split('?')
+    const newSource = source.split('?')
```
[the remark this answers is not in the document]
Comment (EECOLOR) [reply library/eslint-resolver-plugins/absolute-path-resolver-plugin.js:7]: Very nice! That was an accidental edit!

**Comment by EECOLOR** [comment]
@peeke I forwarded the Slack conversation about this in an e-mail. That should give you some background info and also provide a lot of alternative names we considered.

**Comment by peeke** [comment]
I've read the discussion, which greatly clarifies the PR, thanks.

I'm not sold on the name `sub`, because it's not meaningful. It's a subdirectory for the purpose of having a subdirectory. 

My first inituition was `feature`, which I saw mentioned in the discussion as well. `feature` was discarded based on the fact that there was a folder named `phase`, which was a page, not a feature. I think this is a misinterpretation though. It doesn't mean `phase` is a feature, is rather a folder that hold a group of features, for which `phase` is the most descriptive name of what the features inside that group have in common. The term feature works perfectly fine, combined with subfolders based on e.g.:
- functionality (`form`)
- page (`phase`)
- domain ('shopping-cart') 
- flow (`checkout/confirm-order`). 

Anything goes really, as long as it properly describes the features inside. There must be features inside, otherwise you're dealing with machinery or a page as whole.

**Comment by EECOLOR** [comment]
@peeke you have convinced me. @JoostKiens what about you?

And if we choose `feature`, I think I would go for `features` because we also have `pages`.

**Comment by JoostKiens** [comment]
I agree with @peeke's arguments against `sub` and against our reason for discarding `feature`, but aren't `pages` and the contents of `machinery` also `features`?

**Comment by peeke** [comment]
I think they technically are, but you both mentioned good reasons why they should deserve their own folder. Based on those, it'd be easy enough to place them in one of the 3 folders:

- is it a page? -> pages
- is it a generic piece of machinery, that makes sense on its own outside of a specific feature group? -> machinery
- otherwise -> features

**Comment by EECOLOR** [comment]
@JoostKiens ?

**Comment by JoostKiens** [comment]
Still think that the term "features" doesn't distinguish between "features", "pages" and "machinery"

- "pages" are pages, this works (but are also features)
- "machinery" consists of stuff which is reusable across projects, this is a convention we are all used to( but are also features)
- "features" consists of stuff we use **on pages**, but the name doesn't imply this. And what's more, the term "features" doesn't exclude things from "machinery" and "pages", since they all consist of "features".

Ideally we should come up with a term that implies 2 things
1. It's only used "on pages"; things here are never used on the highest or second-highest level of the component tree
2. It doesn't try to say anything about the type of thing: it could be a component, a hook and possibly something else

"Features" works for 2.), but not 1.). 

Right now we use "components" which doesn't work for either, so it is an improvement.

Otoh, it's just a convention, the name doesn't impact the quality of the codebase. Next.js and gatsby use components and they are doing just fine. So 🤷‍♂️

**Comment by EECOLOR** [comment]
Note to self: when we spoke about `sub` in a real life project Joost and I concluded that `features` was a bit better.

**Comment by EECOLOR** [comment]
Created a branch with a better name, will open a pull request there.

**Pull request closed without merging**